### DISCUSSION TITLE
[java] PMD Default Cyclomatic Complexity

### DISCUSSION THREAD

**Discussion by arpitx165** [pmd/pmd#3042] [Q&A] [answered]
**Affects PMD Version: 6.2.0** 

**Description**: PMD shows default Cyclomatic Complexity at class level is 80 and Method level 10.How its computing Class level complexity from methods complexity ?
Edit : I found that it takes sum of all methods complexity but in older function it was taking the average of all methods complexity .I want to know How it is better measurement than older one ?

**Running PMD through:** *[CLI | Maven ]*

**Comment by jsotuyod** [comment, marked as answer]
@arpitx165 the new rule is a more standard measure of Cyclomatic Complexity (more details on how it's computed and sources for such definitions can be found [here](https://pmd.github.io/pmd-6.1.0/pmd_java_metrics_index.html#cyclomatic-complexity-cyclo)).

As for classes, the sum better describes "how complex" a class is, than an average. The old deprecated rule, by averaging, was forgiving of classes that had a bunch of complex methods, as long as it had more simpler methods (ie: getters and setters). The new one, is far more accurate. A class with too many methods, even if simple ones, is large and therefore complex.

**Comment by arpitx165** [comment]
@jsotuyod Okay I got your point but when we will do sum of methods complexity.There can be two scenarios suppose one class which contains 10 methods each has complexity 1 because most are getter and setter methods.So according to Current Rule Complexity will be 10 and In another class there is 1 method which has Complexity 10 So according  to Current rule complexity will be 10.

In both scenarios complexity is 10. What do you think about  complexity in both scenarios ? Is it better measurement ?

**Comment by jsotuyod** [comment]
@arpitx165 I personally don't care about cyclomatic complexity at all. This, along other metrics, have been found to be highly correlated with code size ("The confounding effect of class size on the validity of 
Object-Oriented Metrics" - El Emam et al (2001)). That is, larger classes tend to have more methods, and those methods tend to  be larger and more complex. 

I personally prefer to use the method length / class length rules rather than these. 

However, the sum of complexities tends to be better aligned with code length except on extreme cases (DTOs / java beans), so if I had to choose one I'd probably stick to the new one and ignore those packages.

**Comment by arpitx165** [comment]
@jsotuyod  I found the same we can not stick on only Cyclomatic Complexity we have to consider other parameters btw Thanks for Your time :)
